src/Owin.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.Endpoints.cs

```diff
@@ -1729,6 +1729,81 @@ internal partial class OpenIdConnectServerHandler : AuthenticationHandler<OpenId
+                      !string.Equals(identifier, clientNotification.ClientId)))) {
+                    // If the caller was not authenticated
+                    //  AND the token was issued to an authenticated client
+                    // OR if the caller was authenticated 
+                    //  AND either the token was not issued to an authenticated client OR the token was issued to a different authenticated client
+                    // Then the token belongs to a different client, reject the request.
+                    Options.Logger.WriteVerbose("refresh token issued to different client");
+
+                    await SendPayloadAsync(new JObject {
+                        [OpenIdConnectConstants.Claims.Active] = false
+                    });
+
+                    return;
+                }
+            }
+
+            else if (string.Equals(usage, OpenIdConnectConstants.Usages.IdToken, StringComparison.Ordinal)) {
+                //For Identity tokens
+                if (clientNotification.IsSkipped ||
+                    (!string.IsNullOrWhiteSpace(clientNotification.ClientId) &&
+                     !audiences.Contains(clientNotification.ClientId))) {
+                    // If the caller is not authenticated OR is not one of the specified audiences,
+                    if (ticket.ContainsProperty(OpenIdConnectConstants.Extra.ClientAuthenticated)) {
+                        // And if the token was issued to an authenticated client, it was issued to a different client
+                        Options.Logger.WriteVerbose("identity token issued to different client");
+
+                        await SendPayloadAsync(new JObject {
+                            [OpenIdConnectConstants.Claims.Active] = false
+                        });
+
+                        return;
+                    }
+                }
+            }
+
+            else if (string.Equals(usage, OpenIdConnectConstants.Usages.AccessToken, StringComparison.Ordinal)) {
+                // For access tokens
+                if (!clientNotification.IsSkipped && !string.IsNullOrWhiteSpace(clientNotification.ClientId)) {
+                    if (!audiences.Contains(clientNotification.ClientId)) {
+                        //If the client is authenticated but not in the audiences, they're not allowed to validate the token.
+                        Options.Logger.WriteVerbose("invalid audience for access token");
+
+                        await SendPayloadAsync(new JObject {
+                            [OpenIdConnectConstants.Claims.Active] = false
+                        });
+
+                        return;
+                    }
+                }
+
+                else {
+                    //If the caller is not authenticated
+                    if (!ticket.ContainsProperty(OpenIdConnectConstants.Extra.ClientAuthenticated) && !ticket.Identity.HasClaim(
```

Comment:
I didn't see that overload when I was trying to determine if it had the claim, otherwise I would have used it. My bad. I'll fix this when I get a chance.